File: src/lib/integrations/github.ts
import type { Prisma } from "@/generated/prisma/client";
import { db } from "@/lib/db";
import type { GitHubIntegrationMetadata } from "@/lib/types";

const GITHUB_API = "https://api.github.com";

export type GitHubComment = {
  id: number;
  user: { login: string } | null;
  body: string;
  created_at: string;
};

export type GitHubIssue = {
  number: number;
  title: string;
  body: string | null;
  html_url: string;
  user: { login: string } | null;
  labels: Array<{ name: string }>;
  assignees: Array<{ login: string }>;
  pull_request?: { url: string };
};

export async function getGitHubToken(userId: string): Promise<string | null> {
  const integration = await db.integration.findUnique({
    where: { userId_provider: { userId, provider: "GITHUB" } },
  });
  return integration?.accessToken ?? null;
}

export async function getGitHubMetadata(
  userId: string
): Promise<GitHubIntegrationMetadata> {
  const integration = await db.integration.findUnique({
    where: { userId_provider: { userId, provider: "GITHUB" } },
  });
  return (integration?.metadata as GitHubIntegrationMetadata) ?? {};
}

export async function isGitHubConnected(userId: string) {
  const token = await getGitHubToken(userId);
  return !!token;
}

export async function saveGitHubIntegration(
  userId: string,
  accessToken: string,
  metadata: GitHubIntegrationMetadata
) {
  await db.integration.upsert({
    where: { userId_provider: { userId, provider: "GITHUB" } },
    create: {
      userId,
      provider: "GITHUB",
      accessToken,
      metadata: {
        autoAssign: true,
        autoMention: true,
        autoReview: true,
        autoAckMention: true,
        ...metadata,
      } as Prisma.InputJsonValue,
    },
    update: {
      accessToken,
      metadata: {
        autoAssign: true,
        autoMention: true,
        autoReview: true,
        autoAckMention: true,
        ...metadata,
      } as Prisma.InputJsonValue,
    },
  });
}

export async function updateGitHubSettings(
  userId: string,
  settings: Partial<GitHubIntegrationMetadata>
) {
  const existing = await getGitHubMetadata(userId);
  await db.integration.update({
    where: { userId_provider: { userId, provider: "GITHUB" } },
    data: {
      metadata: { ...existing, ...settings } as Prisma.InputJsonValue,
    },
  });
}

export async function githubFetch<T>(
  token: string,
  path: string,
  init?: RequestInit
): Promise<T> {
  const res = await fetch(`${GITHUB_API}${path}`, {
    ...init,
    headers: {
      Accept: "application/vnd.github+json",
      Authorization: `Bearer ${token}`,
      "X-GitHub-Api-Version": "2022-11-28",
      ...init?.headers,
    },
  });

  if (!res.ok) {
    throw new Error(`GitHub API error: ${res.status} ${await res.text()}`);
  }

  return res.json() as Promise<T>;
}

export async function fetchGitHubUser(token: string) {
  return githubFetch<{ login: string; id: number; avatar_url: string }>(
    token,
    "/user"
  );
}

export async function fetchIssueOrPull(
  token: string,
  repo: string,
  number: number
): Promise<GitHubIssue> {
  return githubFetch<GitHubIssue>(token, `/repos/${repo}/issues/${number}`);
}

export async function fetchIssueComments(
  token: string,
  repo: string,
  number: number
): Promise<GitHubComment[]> {
  return githubFetch<GitHubComment[]>(
    token,
    `/repos/${repo}/issues/${number}/comments`
  );
}

export async function postIssueComment(
  userId: string,
  repo: string,
  issueNumber: number,
  body: string
) {
  const token = await getGitHubToken(userId);
  if (!token) throw new Error("GitHub not connected");

  return githubFetch<{ id: number; html_url: string }>(
    token,
    `/repos/${repo}/issues/${issueNumber}/comments`,
    {
      method: "POST",
      body: JSON.stringify({ body }),
    }
  );
}

export async function addIssueLabels(
  userId: string,
  repo: string,
  issueNumber: number,
  labels: string[]
) {
  const token = await getGitHubToken(userId);
  if (!token) throw new Error("GitHub not connected");

  return githubFetch<unknown>(
    token,
    `/repos/${repo}/issues/${issueNumber}/labels`,
    {
      method: "POST",
      body: JSON.stringify({ labels }),
    }
  );
}

export async function findUserByGitHubLogin(login: string) {
  const integrations = await db.integration.findMany({
    where: { provider: "GITHUB" },
  });

  for (const integration of integrations) {
    const meta = integration.metadata as GitHubIntegrationMetadata;
    if (meta.githubLogin?.toLowerCase() === login.toLowerCase()) {
      return integration.userId;
    }
  }

  return null;
}

export function externalIdForIssue(repo: string, number: number) {
  return `${repo}#${number}`;
}

File: backend/app/services/integrations/github.py
import secrets
from typing import Any

import httpx
from sqlalchemy import select

from app.database import AsyncSessionLocal
from app.models import Integration, IntegrationProvider


GITHUB_API = "https://api.github.com"


def new_id() -> str:
    return secrets.token_hex(12)


async def get_github_token(user_id: str) -> str | None:
    async with AsyncSessionLocal() as db:
        result = await db.execute(
            select(Integration).where(
                Integration.userId == user_id,
                Integration.provider == IntegrationProvider.GITHUB,
            )
        )
        integration = result.scalar_one_or_none()
    return integration.accessToken if integration else None


async def get_github_metadata(user_id: str) -> dict[str, Any]:
    async with AsyncSessionLocal() as db:
        result = await db.execute(
            select(Integration).where(
                Integration.userId == user_id,
                Integration.provider == IntegrationProvider.GITHUB,
            )
        )
        integration = result.scalar_one_or_none()
    if not integration or not integration.metadata_:
        return {}
    return integration.metadata_


async def is_github_connected(user_id: str) -> bool:
    token = await get_github_token(user_id)
    return bool(token)


async def save_github_integration(
    user_id: str,
    access_token: str,
    metadata: dict[str, Any] | None = None,
) -> None:
    meta = {
        "autoAssign": True,
        "autoMention": True,
        "autoReview": True,
        "autoAckMention": True,
        **(metadata or {}),
    }

    async with AsyncSessionLocal() as db:
        result = await db.execute(
            select(Integration).where(
                Integration.userId == user_id,
                Integration.provider == IntegrationProvider.GITHUB,
            )
        )
        integration = result.scalar_one_or_none()

        if integration:
            integration.accessToken = access_token
            integration.metadata_ = meta
        else:
            db.add(
                Integration(
                    id=new_id(),
                    userId=user_id,
                    provider=IntegrationProvider.GITHUB,
                    accessToken=access_token,
                    metadata_=meta,
                )
            )
        await db.commit()


async def update_github_settings(
    user_id: str,
    settings: dict[str, Any],
) -> None:
    existing = await get_github_metadata(user_id)
    async with AsyncSessionLocal() as db:
        result = await db.execute(
            select(Integration).where(
                Integration.userId == user_id,
                Integration.provider == IntegrationProvider.GITHUB,
            )
        )
        integration = result.scalar_one_or_none()
        if integration:
            integration.metadata_ = {**existing, **settings}
            await db.commit()


async def github_fetch(
    token: str,
    path: str,
    method: str = "GET",
    json_body: dict[str, Any] | None = None,
) -> Any:
    headers = {
        "Accept": "application/vnd.github+json",
        "Authorization": f"Bearer {token}",
        "X-GitHub-Api-Version": "2022-11-28",
    }
    async with httpx.AsyncClient() as client:
        response = await client.request(
            method,
            f"{GITHUB_API}{path}",
            headers=headers,
            json=json_body,
        )
        if not response.is_success:
            raise RuntimeError(f"GitHub API error: {response.status_code} {response.text}")
        return response.json()


async def fetch_github_user(token: str) -> dict[str, Any]:
    return await github_fetch(token, "/user")


async def fetch_issue_or_pull(token: str, repo: str, number: int) -> dict[str, Any]:
    return await github_fetch(token, f"/repos/{repo}/issues/{number}")


async def fetch_issue_comments(token: str, repo: str, number: int) -> list[dict[str, Any]]:
    return await github_fetch(token, f"/repos/{repo}/issues/{number}/comments")


async def post_issue_comment(
    user_id: str,
    repo: str,
    issue_number: int,
    body: str,
) -> dict[str, Any]:
    token = await get_github_token(user_id)
    if not token:
        raise RuntimeError("GitHub not connected")
    return await github_fetch(
        token,
        f"/repos/{repo}/issues/{issue_number}/comments",
        method="POST",
        json_body={"body": body},
    )


async def add_issue_labels(
    user_id: str,
    repo: str,
    issue_number: int,
    labels: list[str],
) -> Any:
    token = await get_github_token(user_id)
    if not token:
        raise RuntimeError("GitHub not connected")
    return await github_fetch(
        token,
        f"/repos/{repo}/issues/{issue_number}/labels",
        method="POST",
        json_body={"labels": labels},
    )


async def find_user_by_github_login(login: str) -> str | None:
    async with AsyncSessionLocal() as db:
        result = await db.execute(
            select(Integration).where(Integration.provider == IntegrationProvider.GITHUB)
        )
        integrations = result.scalars().all()

    login_lower = login.lower()
    for integration in integrations:
        meta = integration.metadata_ or {}
        github_login = meta.get("githubLogin")
        if github_login and github_login.lower() == login_lower:
            return integration.userId
    return None


def external_id_for_issue(repo: str, number: int) -> str:
    return f"{repo}#{number}"
